### PR TITLE
Make loading snapshots parallel

### PR DESCRIPTION
**Pull request by aawsome**
What does this PR change? What problem does it solve?
-----------------------------------------------------

Loads snaphots files in parallel, like already done by `restic check`.
This speeds up many commands if there is a large number of snapshots in the repository.

Note that this should be also included in #3121 (either here or there depending on the order of PR merge)

Was the change discussed in an issue or in the forum before?
------------------------------------------------------------

The issue came up quite often, but so far only the possibility to change the repo format (snapshot blobs) has been discussed.

Checklist
---------

- [x] I have read the [Contribution Guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches)
- [x] I have enabled [maintainer edits for this PR](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork)
-  I have not added tests for all changes in this PR, but existing tests still pass
-  I have not added documentation for the changes (in the manual)
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (template [here](https://github.com/restic/restic/blob/master/changelog/TEMPLATE))
- [x] I have run `gofmt` on the code in all commits
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits)
- [x] I'm done, this Pull Request is ready for review

### CODE REVIEW

**Review by fd0 — CHANGES_REQUESTED** [review]

File: internal/restic/snapshot.go [hunk 1]
```diff
@@ -66,14 +69,68 @@ func LoadSnapshot(ctx context.Context, repo Repository, id ID) (*Snapshot, error
...
+			case <-ctx.Done():
+				return nil
+			case ch <- id:
+			}
 			return nil
+		})
+	})
+
+	// a worker receives an index ID from ch, loads the snapshot and the tree,
+	// and adds the result to errs and trees.
```
[the remark this answers is not in the document]
Comment (aawsome) [reply internal/restic/snapshot.go:104]: Ups, copy-past error..

File: internal/restic/snapshot_find.go [hunk 2]
```diff
@@ -29,14 +29,13 @@ func FindLatestSnapshot(ctx context.Context, repo Repository, targets []string,
...
+		latestID *ID
 		found    bool
 	)
 
-	err = repo.List(ctx, SnapshotFile, func(snapshotID ID, size int64) error {
-		snapshot, err := LoadSnapshot(ctx, repo, snapshotID)
+	err = ForAllSnapshots(ctx, repo, nil, func(snapshot *Snapshot, err error) error {
 		if err != nil {
-			return errors.Errorf("Error loading snapshot %v: %v", snapshotID.Str(), err)
+			return errors.Errorf("Error loading snapshot %v: %v", snapshot.ID().Str(), err)
```
[the remark this answers is not in the document]
Comment (aawsome) [reply internal/restic/snapshot_find.go]: You are right. Is changed now.

**Comment by MichaelEischer** [comment]
Hmm, this reminds me of https://github.com/MichaelEischer/restic/commit/3da6b9de9d51290a868c034418f662cf8d6a2250 from my https://github.com/MichaelEischer/restic/commits/streamtrees branch. From what I see the main difference is the function interface `func StreamSnapshots(ctx context.Context, wg *errgroup.Group, repo restic.Repository, excludeIDs restic.IDSet) <-chan SnapshotItem`. The main difference is that I've used a channel instead of a callback (the parameters themselves are equivalent). Using a channel avoids using a mutex but requires a bit more ceremony, see https://github.com/MichaelEischer/restic/commit/7c68668ef76bde63347a2094a8d31bbbd2eecfaf . @fd0 do you have a preference?

I'd also prefer to not add additional code to the restic package, but I'm not sure whether a new "parallel" package would be a solution.

**Comment by aawsome** [comment]
@MichaelEischer Thanks for the hint - I didn't fully remember what was included in your change.

Besides the decision about how to handle this (I'm open to both possibilities) we should add this parallel loading of snapshots not only to the processing of all tress. All snapshots are also read is for instance also used when searching for the latest snapshot which is used in several places. This is why I proposed to change it within `internal/restic`.

**Comment by fd0** [comment]
@MichaelEischer I think I prefer the callback variant. This allows easier error handling IMHO. :)

**Comment by MichaelEischer** [comment]
@aawsome Most snapshot users use the `Find*Snapshot` functions, see https://github.com/MichaelEischer/restic/commit/44dec3328f17de18764d12d9a77b7fb1b5f9a925 . `LoadAllSnapshots` is only used by prune (see https://github.com/MichaelEischer/restic/commit/1bd0370e6e34b9ab8662eb9126fc67486fffe8f7), that change is relatively independent of my StreamTrees PR (well, the merge conflict there should be rather easy to resolve).

We could of course move the function to find snapshots to a different package. That way we would limit the restic package to fundamental data structures and put other helper somewhere else.

File: internal/restic/snapshot.go [hunk 3]
```diff
@@ -66,14 +69,67 @@ func LoadSnapshot(ctx context.Context, repo Repository, id ID) (*Snapshot, error
 	return sn, nil
 }
 
-// LoadAllSnapshots returns a list of all snapshots in the repo.
-// If a snapshot ID is in excludeIDs, it will not be included in the result.
-func LoadAllSnapshots(ctx context.Context, repo Repository, excludeIDs IDSet) (snapshots []*Snapshot, err error) {
-	err = repo.List(ctx, SnapshotFile, func(id ID, size int64) error {
-		if excludeIDs.Has(id) {
+const defaultParallelism = 5
```
Comment (MichaelEischer) [inline internal/restic/snapshot.go:72]: I'd prefer to call that constant `loadSnapshotParallelism`.

**Review by MichaelEischer — COMMENTED** [review]
@aawsome How should we proceed with this PR? Do you want to port the change to the debug command from https://github.com/MichaelEischer/restic/commit/44dec3328f17de18764d12d9a77b7fb1b5f9a925#diff-0f07299f9156bf36536e4748f00ebe242ca7d360750d56cba41219fdbf4e372a and the replacement of LoadAllSnapshots from https://github.com/MichaelEischer/restic/commit/1bd0370e6e34b9ab8662eb9126fc67486fffe8f7 (which has the benefit that it doesn't have to keep all snapshots in memory) to your `ForAllSnapshots` variant? Or should I do that in a separate PR?

This PR also needs a changelog entry IMHO. After all `prune` and every command which has to search for a snapshot is faster now.

**Comment by aawsome** [comment]
@MichaelEischer Thanks for asking. I was in fact waiting for directions and just realized that I didn't understand the hint in your comment last week. Thanks for clarifying your proposal! I now changed `debug` and `prune` as you suggested (the parallelized tree walk needs to be adjusted in your tree PR).

I think this is now ready for merging and IMO should be merged. Sorry again for "stealing" your work, I didn't realize that you were already working on this topic...

File: internal/restic/snapshot.go (hunk 3 above)
Comment (aawsome) [reply internal/restic/snapshot.go]: is changed.

**Review by MichaelEischer — APPROVED** [review]
LGTM. I've tweaked the changelog entry a bit.

**Review by fd0 — COMMENTED** [review]
LGTM!

**Pull request merged**